File: app/src/main/java/io/mgba/UI/Fragments/Main/GameFragment.java
package io.mgba.UI.Fragments.Main;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import org.lucasr.twowayview.layout.TwoWayView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.mgba.Adapters.GameAdapter;
import io.mgba.Presenter.GamesPresenter;
import io.mgba.Presenter.Interfaces.IGamesPresenter;
import io.mgba.Data.Database.Game;
import io.mgba.R;
import io.mgba.UI.Fragments.Interfaces.IGamesFragment;

public class GameFragment extends Fragment implements SwipeRefreshLayout.OnRefreshListener, IGamesFragment<Game> {

    private static final String TAG = "BaseFragment";

    @BindView(R.id.no_content_container)
    protected RelativeLayout mNoContentView;
    @BindView(R.id.content_recyclerView)
    protected TwoWayView mRecyclerView;
    @BindView(R.id.no_content_image)
    protected ImageView noContentImage;
    @BindView(R.id.no_content_message)
    protected TextView noContentMessage;
    @BindView(R.id.swipeRefreshLayout)
    SwipeRefreshLayout mSwipeRefreshLayout;

    private GameAdapter adapter;
    protected IGamesPresenter controller;
    protected View mView;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        controller = new GamesPresenter(this, this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mView = prepareView(inflater, container, savedInstanceState);
        ButterKnife.bind(this, mView);

        prepareDrawables();
        prepareRecyclerView();

        showContent(false);
        loadGames();

        return mView;
    }

    protected void prepareRecyclerView() {
        mSwipeRefreshLayout.setOnRefreshListener(this);
        mSwipeRefreshLayout.setColorSchemeColors(getResources().getColor(R.color.pink_accent_color),
                                                 getResources().getColor(R.color.colorPrimary),
                                                 getResources().getColor(R.color.green_accent_color),
                                                 getResources().getColor(R.color.yellow_accent_color),
                                                 getResources().getColor(R.color.cyan_accent_color));
        mRecyclerView.setHasFixedSize(true);
        adapter = new GameAdapter(this, getContext(), controller.getOnClick(), mRecyclerView);
        mRecyclerView.setAdapter(adapter);

    }

    protected void prepareDrawables() {
        noContentImage.setImageDrawable(getResources().getDrawable(R.drawable.ic_videogame_asset_grey_500_48dp));
        noContentMessage.setText(R.string.no_games);
    }

    @Override
    public void showContent(boolean state) {
        mNoContentView.setVisibility(state ? View.GONE : View.VISIBLE);
        mRecyclerView.setVisibility(state ? View.VISIBLE : View.GONE);
    }

    protected View prepareView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.content_fragment, container, false);
    }

    protected void loadGames() {
        controller.loadGames(getILibrary());
    }

    private io.mgba.UI.Activities.Interfaces.ILibrary getILibrary(){
        return ((io.mgba.UI.Activities.Interfaces.ILibrary)getActivity());
    }

    @Override
    public void onRefresh() {
        controller.onRefresh(getILibrary());
    }

    @Override
    public void onStop() {
        super.onStop();
        controller.onDestroy();
    }

    @Override
    public void swapContent(List<Game> items) {
        adapter.swap(items);
    }

    @Override
    public void stopRefreshing() {
        mSwipeRefreshLayout.setRefreshing(false);
    }


    @Override
    public void handleItemClick(Game game) {
        getILibrary().showBottomSheet(game);
    }
}

File: app/src/main/java/io/mgba/ui/fragments/main/GameFragment.kt
package io.mgba.ui.fragments.main

import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.NonNull
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import io.mgba.adapters.GameAdapter
import io.mgba.Constants
import io.mgba.presenter.GamesPresenter
import io.mgba.presenter.interfaces.IGamesPresenter
import io.mgba.data.database.model.Game
import io.mgba.R
import io.mgba.ui.fragments.interfaces.IGamesFragment
import io.mgba.ui.views.GameInformationView
import io.mgba.mgba
import kotlinx.android.synthetic.main.games_list_view.*

open class GameFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener, IGamesFragment<Game> {
    private lateinit var adapter: GameAdapter
    private lateinit var controller: IGamesPresenter

    private val iLibrary: io.mgba.ui.activities.interfaces.ILibrary
        get() = activity as io.mgba.ui.activities.interfaces.ILibrary


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return prepareView(inflater, container)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if(arguments == null)
            mgba.report(Exception("Game Fragment creation failed because args are null"))

        val platform = arguments!!.getInt(Constants.ARG_PLATFORM)
        controller = GamesPresenter(platform, this)

        prepareDrawables()
        prepareRecyclerView()

        if (savedInstanceState != null) {
            list.layoutManager!!
                    .onRestoreInstanceState(savedInstanceState.getParcelable<Parcelable>(Constants.MAIN_RECYCLER_CONTENT))
        }

        showContent(false)
        loadGames()
    }

    private fun prepareRecyclerView() {
/*        main_refresh.setOnRefreshListener(this)
        main_refresh.setColorSchemeColors(ContextCompat.getColor(context!!, R.color.pink_accent_color),
                ContextCompat.getColor(context!!, R.color.colorPrimary),
                ContextCompat.getColor(context!!, R.color.green_accent_color),
                ContextCompat.getColor(context!!, R.color.yellow_accent_color),
                ContextCompat.getColor(context!!, R.color.cyan_accent_color))*/
        list.setHasFixedSize(true)
        adapter = GameAdapter(this, context!!, { controller.onClick(it) }, list)
        list.adapter = adapter

    }

    protected open fun prepareDrawables() {
        noContentIcon.setImageDrawable(ContextCompat.getDrawable(context!!, R.drawable.ic_videogame_asset_grey_500_48dp))
        noContentMessage.setText(R.string.no_games)
    }

    override fun showContent(state: Boolean) {
        noContent.visibility = if (state) View.GONE else View.VISIBLE
        list.visibility = if (state) View.VISIBLE else View.GONE
    }

    private fun prepareView(inflater: LayoutInflater?, container: ViewGroup?): View {
        return inflater!!.inflate(R.layout.games_list_view, container, false)
    }

    private fun loadGames() {
        controller.loadGames(iLibrary)
    }

    override fun onRefresh() {
        controller.onRefresh(iLibrary)
    }

    override fun onStop() {
        super.onStop()
        controller.onDestroy()
    }

    override fun swapContent(items: List<Game>) {
        showContent(items.isNotEmpty())
        adapter.swap(items)
    }

    override fun stopRefreshing() {
        refreshLayout.isRefreshing = false
    }


    override fun handleItemClick(game: Game) {
        val sheet = GameInformationView.show(game)
        sheet.show(fragmentManager, TAG + "_SHEET")
    }

    override fun onSaveInstanceState(@NonNull outState: Bundle) {
        super.onSaveInstanceState(outState)
        list.layoutManager?.let {
            outState.putParcelable(Constants.MAIN_RECYCLER_CONTENT, it.onSaveInstanceState())
        }
    }

    companion object {
        private const val TAG = "BaseFragment"
    }
}
